File: dsa-05-main/cpp/day07/doubly_list.cpp
#include <iostream>
using namespace std;

class doubly_list; // forward declaration class

class node {
private:
	int data;
	node* next;
	node* prev;
public:
	node() {
		data = 0;
		next = NULL;
		prev = NULL;
	}
	node(int val) {
		data = val;
		next = NULL;
		prev = NULL;
	}
	friend class doubly_list;
};

class doubly_list {
private:
	node* head;
public:
	doubly_list() {
		head = NULL;
	}
	void display() {
		cout << "FWD LIST: ";
		node* trav = head;
		while (trav != NULL) {
			cout << trav->data << ", ";
			trav = trav->next;
		}
		cout << endl;
	}
	void display_rev() {
		cout << "REV LIST: ";
		// traverse till last node
		node* trav;
		if (head != NULL) {
			trav = head;
			while (trav->next != NULL)
				trav = trav->next;
			// visit each node in reverse order (using prev pointer)
			while (trav != NULL) {
				cout << trav->data << ", ";
				trav = trav->prev;
			}
		}
		cout << endl;
	}
	void add_first(int val) {

	}
	void add_last(int val) {
		node* newnode, * trav;
		// alloc new node and init it
		newnode = new node(val);
		// special case: if list is empty, newnode itself is first node
		if (head == NULL)
			head = newnode;
		else {
			// traverse till last node
			trav = head;
			while (trav->next != NULL)
				trav = trav->next;
			// add new node to the next of last node
			trav->next = newnode;
			// newnode prev to last node
			newnode->prev = trav;
		}
	}
	void add_at_pos(int pos, int val) {

	}
	void del_first() {
	}

	void del_at_pos(int pos) {
	}

	~doubly_list() {
		del_all();
	}

	void del_all() {
		while (head != NULL)
			del_first();
	}
};

int main() {
	doubly_list l1;
	l1.add_last(10);
	l1.add_last(20);
	l1.add_last(30);
	l1.add_first(40);
	l1.add_at_pos(3, 50);
	l1.display();
	l1.add_at_pos(1, 60);
	l1.display();
	l1.add_at_pos(7, 70);
	l1.display();
	l1.del_first();
	l1.display();
	l1.del_first();
	l1.display();
	l1.del_at_pos(3);
	l1.display();
	l1.del_at_pos(4);
	l1.display();
	l1.del_at_pos(5);
	l1.display();
	l1.del_all();
	l1.display();
	l1.display_rev();
	return 0;
}

File: dsa-05-main/python/day07/doubly_list.py
class Node:
	def __init__(self, val=0):
		self.data = val
		self.next = None
		self.prev = None


class DoublyList:

	def __init__(self):
		self.head = None
		self.count = 0

	# time complexity - O(1)
	def addFirst(self, val):
		newNode = Node(val)
		if self.head is None:
			self.head = newNode
		else:
			newNode.next = self.head
			self.head.prev = newNode
			self.head = newNode
		self.count = self.count + 1

	# time complexity - O(n)
	def addLast(self, val):
		newNode = Node(val)
		if self.head is None:
			self.head = newNode
		else:
			trav = self.head
			while trav.next is not None:
				trav = trav.next
			trav.next = newNode
			newNode.prev = trav
		self.count = self.count + 1

	# time complexity - O(n)
	def display(self):
		temp = None
		# display forward
		print("Fwd List: ", end='')
		trav = self.head
		while trav is not None:
			print(trav.data, end=", ")
			temp = trav
			trav = trav.next
		print()
		# display reverse
		print("Rev List: ", end='')
		trav = temp
		while trav is not None:
			print(trav.data, end=", ")
			trav = trav.prev
		print()

	# time complexity - O(n)
	def addAtPos(self, val, pos):
		# // spl3: add at last pos
		if pos == self.count+1:
			self.addLast(val)
		# // spl4: invalid pos
		elif pos < 1 or pos > self.count+1:
			raise Exception("Invalid position " + str(pos))
		# // spl1: list is empty or spl2: pos is 1
		elif self.head is None or pos == 1:
			self.addFirst(val)
		else:
			newNode = Node(val)
			# // traverse till pos-1
			trav = self.head
			for i in range(1, pos-1):
				trav = trav.next
			# // mark its next node as temp
			temp = trav.next
			# // add newNode between trav & temp
			newNode.next = temp
			newNode.prev = trav
			trav.next = newNode
			temp.prev = newNode
			self.count = self.count + 1

	# time complexity - O(1)
	def delFirst(self):
		# // spl1: list is empty
		if self.head is None:
			raise Exception("List is empty.")
		# // spl2: if list have only one element
		if self.head.next is None:
			self.head = None
		else:
			self.head = self.head.next
			self.head.prev = None
		self.count = self.count - 1

	def delAtPos(self, pos):
		# // spl1: list is empty or spl2: pos = 1
		if self.head is None or pos == 1:
			self.delFirst()
		else:
			# // spl3: invalid pos
			if pos < 1 or pos > self.count:
				raise Exception("Invalid position " + str(pos))
			# // traverse till pos
			trav = self.head
			for i in range(1, pos):
				trav = trav.next
			# // delete trav node
			trav.prev.next = trav.next
			if(trav.next is not None): # // spl4: if not last node to delete.
				trav.next.prev = trav.prev
			self.count = self.count - 1

	# time complexity - O(n)
	def delAll(self):
		while self.head is not None:
			self.delFirst()
		# head = None

	def countNodes(self):
		cnt = 0
		trav = self.head
		while trav is not None:
			cnt = cnt + 1
			trav = trav.next
		return cnt


def main():
	l1 = DoublyList()
	l1.addLast(11)
	l1.addLast(22)
	l1.addLast(33)
	l1.addFirst(44)
	l1.display()
	l1.addAtPos(55, 5)
	l1.display()
	l1.delFirst()
	l1.display()
	l1.delAtPos(2)
	l1.display()


if __name__ == "__main__":
	main()

